### PR TITLE
feat: Add single select and inverse selection to numeric range (#16722)

### PR DESCRIPTION
**Pull request by mhoustonataegis**
### SUMMARY
Support the ability to filter by numeric range as a single maximum, minimum, or exact value.

### Filters config form BEFORE:
Before_1

### Range filter plugin BEFORE:
Before_2

### Filters config form AFTER:
After_1

### Range filter plugin AFTER, using MAXIMUM value:
After_2

### Range filter plugin AFTER, using EXACT value:
After_3

### Range filter plugin AFTER, using MINIMUM value:
After_4

### TESTING INSTRUCTIONS
Using a numerical range filter, choose "Single Value" in the advanced section.  This will reveal 3 single value type options supporting the following behaviors:
	* Minimum: Filter values greater than or equal to the selected value on the range slider.
	* Exact: Filter values equal to the selected value on the range slider.
	* Maximum: Filter values less than or equal to the selected value on the range slider.
	
If "Single Value" is not checked, the previous default range slider behavior applies.
If "Single Value" is checked and Minimum is chosen, the slider should support a single thumb sliding from the right hand side of the control.
If "Single Value" is checked and Exact is chosen, the slider should support a single thumb sliding in the middle of the control with no highlighted range displayed.
If "Single Value" is checked and Maximum is chosen, the slider should support a single thumb sliding from the left hand side of the control.

### ADDITIONAL INFORMATION

- [x] Has associated issue: #16722
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/filters/components/Range/SingleValueType.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export enum SingleValueType {
```
Comment (geido) [inline superset-frontend/src/filters/components/Range/SingleValueType.ts:1]: @mhoustonataegis this file is missing the license. CI won't pass without it

**Comment by BillGreerUX** [comment]
This PR is exactly what I requested with https://github.com/apache/superset/issues/16722.  This will solve our use case well.

**Comment by riahk** [comment]
Code looks good, but the MIN slider UI is a bit misleading.
Screen Shot 2021-11-23 at 11 43 37 AM
Screen Shot 2021-11-23 at 11 43 43 AM

I get that the `reverse` flag is meant to distinguish it from the MAX slider. I dug around through the antd docs and couldn't find a decent solution, though. I'm wondering if there's a way to style the MIN slider so the colors are inverted, here? That way the blue would appropriately distinguish the included range.
Also — and this is probably excessive, since the user can name the filter whatever they want — maybe having an indicator next to the label of whether the slider is MIN/MAX/EXACT could be useful?

**Comment by villebro** [comment]
> I'm wondering if there's a way to style the MIN slider so the colors are inverted, here? That way the blue would appropriately distinguish the included range.

@riahk I agree, this should be doable fairly easily. To achieve an inverse selection, I simply flipped around the the background colors of `ant-slider-rail` and `ant-slider-track` on the [Slider example page](https://ant.design/components/slider/) and the end result looks pretty ok:

image

Probably requires some additional fine tuning (the hovers need to also be updated), but in the absence of full support for this in the AntD component, this may be the most visually appealing solution for now.

File: superset-frontend/src/filters/components/Range/RangeFilterPlugin.tsx [hunk 2]
```diff
@@ -197,16 +257,53 @@ export default function RangeFilterPlugin(props: PluginFilterRangeProps) {
...
+                max={max}
+                value={minMax[maxIndex]}
+                tipFormatter={tipFormatter}
+                marks={marks}
+                onAfterChange={value => handleAfterChange([min, value])}
+                onChange={value => handleChange([min, value])}
+              />
+            )}
+            {enableSingleMinValue && (
+              <Slider
```
Comment (riahk) [inline superset-frontend/src/filters/components/Range/RangeFilterPlugin.tsx:272]: As a follow-up to my earlier comment about swapping bg colors: I played around with the styles here to give an idea of what that might look like!

```
const StyledMinSlider = styled(Slider)`
  .ant-slider-rail {
    background-color: #99e7f0;
  }

  .ant-slider-track {
    background-color: #f5f5f5;
  }

  &:hover {
    .ant-slider-rail {
      background-color: #99e7f0;
    }

    .ant-slider-track {
      background-color: #f5f5f5;
    }
  }
`;
```

Result:
Screen Shot 2021-11-24 at 10 18 07 AM
Screen Shot 2021-11-24 at 10 18 12 AM

**Review by riahk — COMMENTED** [review]
Is inverse selection actually included in this PR? or is that somewhere else?

**Comment by mhoustonataegis** [comment]
To maintain the visual distinction between a min and max filter, it may make more sense to flip the values on the range rather than flipping the highlighted section.  

For example, there is no visual distinction between a min and max slider if the highlighted section is flipped.  Here, "Months" is a max range and "Years" is a min range, but there is no visual way to discern the difference:

image

However, with the range values flipped, the highlighted section could continue to exist on the right while also corresponding to the correct filtered values:

image

This would result in the visual distinction between a min and max slider:

image

**Comment by mhoustonataegis** [comment]
> Is inverse selection actually included in this PR? or is that somewhere else?

There is no explicit inverse selection supported here.

**Comment by BillGreerUX** [comment]
> This would result in the visual distinction between a min and max slider:
> 
> image

This design is a better user experience.  This visual pattern allows for realization of minimum and maximum. By allowing the visual distinction between min and max, there is not a need to add an indicator next to the label of whether the slider is MIN/MAX/EXACT.  With the visual distinction and not textual, it will be an easier transition when moving too multilingual in future releases. 

Also, the above visual pattern matches the range pattern below.
>   image

I feel @mhoustonataegis suggested implementation would be the best.

**Comment by riahk** [comment]
@mhoustonataegis @BillGreerUX These are excellent points! I like having the colors inverted for that distinction.

What's confusing me still is that it's not clear to me which ends of the scale are high/low. With the `reverse` flag on the min slider, the colors invert, but so do the position of the high/low values! i.e. the "low" value is actually on the right, and the "high" is on the left, which can be confusing.

My suggestion is to remove the `reverse` flag from the min slider, and then add the custom styling mentioned in my earlier comment to invert the colors. The effect is as follows:
Screen Shot 2021-11-29 at 12 39 38 PM

In this case, both sliders have the low value on the left, and the color inversion provides a distinction while still clearly indicating the inclusive range for the filter. Thoughts?

**Comment by mhoustonataegis** [comment]
> @mhoustonataegis @BillGreerUX These are excellent points! I like having the colors inverted for that distinction.
> 
> What's confusing me still is that it's not clear to me which ends of the scale are high/low. With the `reverse` flag on the min slider, the colors invert, but so do the position of the high/low values! i.e. the "low" value is actually on the right, and the "high" is on the left, which can be confusing.
> 
> My suggestion is to remove the `reverse` flag from the min slider, and then add the custom styling mentioned in my earlier comment to invert the colors. The effect is as follows: Screen Shot 2021-11-29 at 12 39 38 PM
> 
> In this case, both sliders have the low value on the left, and the color inversion provides a distinction while still clearly indicating the inclusive range for the filter. Thoughts?

Yes, my suggestion included removing the `reverse` flag, so that both min and max sliders used the same number lines for their values, but only different highlighting based upon your suggested css.  So I think we're on the same page.  I'll push up a new commit.

File: superset-frontend/src/filters/components/Range/RangeFilterPlugin.tsx [hunk 3]
```diff
@@ -32,6 +32,39 @@ import { StatusMessage, StyledFormItem, FilterPluginStyle } from '../common';
...
+}>`
+  ${({ theme, validateStatus }) => `
+  .ant-slider-rail {
+    background-color: ${
+      validateStatus ? theme.colors[validateStatus]?.light1 : LIGHT_BLUE
+    };    
+  }
+
+  .ant-slider-track {
+    background-color: ${LIGHT_GRAY} !important;
```
Comment (riahk) [inline superset-frontend/src/filters/components/Range/RangeFilterPlugin.tsx:51]: are we positive that this `!important` is needed, or are we just being cautious?

**Review by riahk — APPROVED** [review]
LGTM!

**Pull request merged**